compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -709,6 +721,84 @@ fn resolve_trait_impls(
 
     methods
 }
+
+fn check_methods_signatures(
+    resolver: &mut Resolver,
+    impl_methods: &Vec<(FileId, FuncId)>,
+    trait_id: TraitId,
+    errors: &mut Vec<FileDiagnostic>,
+) {
+    let the_trait_shared = resolver.interner.get_trait(trait_id);
+    let the_trait = the_trait_shared.borrow();
+
+    let self_type = resolver.get_self_type().expect("trait impl must have a Self type");
+
+    // Temporarily bind the trait's Self type to self_type so we can type check
+    let _ = the_trait.self_type_typevar.borrow_mut().bind_to(self_type, the_trait.span);
+
+    for (file_id, func_id) in impl_methods {
+        let meta = resolver.interner.function_meta(func_id);
+        let func_name = resolver.interner.function_name(func_id).to_owned();
+
+        let mut typecheck_errors = Vec::new();
+
+        // `method` is none in the case where the impl block has a method that's not part of the Trait
+        if let Some(method) =
+            the_trait.methods.iter().find(|method| method.name.0.contents == func_name)
+        {
+            let function_typ = meta.typ.instantiate(resolver.interner);
+
+            if let Type::Function(params, _, _) = function_typ.0 {
+                if method.arguments.len() == params.len() {
+                    // Check the parameters of the impl method against the parameters of the trait method
+                    for (parameter_index, ((expected, actual), (hir_pattern, _, _))) in
+                        method.arguments.iter().zip(&params).zip(&meta.parameters.0).enumerate()
+                    {
+                        expected.unify(actual, &mut typecheck_errors, || {
+                            TypeCheckError::TraitMethodParameterTypeMismatch {
+                                method_name: func_name.to_string(),
+                                expected_typ: expected.to_string(),
+                                actual_typ: actual.to_string(),
+                                parameter_span: hir_pattern.span(),
+                                parameter_index: parameter_index + 1,
+                            }
+                        });
+                    }
+                } else {
+                    errors.push(
+                        DefCollectorErrorKind::MismatchTraitImplementationNumParameters {
+                            actual_num_parameters: meta.parameters.0.len(),
+                            expected_num_parameters: method.arguments.len(),
+                            trait_name: the_trait.name.to_string(),
+                            method_name: func_name.to_string(),
+                            span: meta.location.span,
+                        }
+                        .into_file_diagnostic(*file_id),
+                    );
+                }
+            }
```

Comment:
Missing an `else` case, I assume it should be `unreachable!()`

Reply:
It's reachable but ignored on purpose. This is the case where the impl block has a method that doesn't match anything in the trait definition. A `MethodNotInTrait` error has already been emitted during collection, but we still go through the motions and type check the signatures, since there could be other impl methods, which do match a trait method for which we could also emit useful errors. e.g:


```rust
trait Asd {
    fn foo() -> i32;
    fn bar() -> u32;
}

impl Asd for Foo {
    // no matching mathod in the trait, ignored during signature checking
    fn fooooo() {} 

    // here it matches the name of `bar` and we emit a TypeMismatch for the return type
    fn bar() -> SomethingBad {...} 
}
```


I've left a comment explaining that: https://github.com/noir-lang/noir/pull/2652/commits/4d9d34566784440a06e3e431bc316dd9ce022fb0